ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import tech.pegasys.teku.core.operationvalidators.OperationStateTransitionValidator;
+import tech.pegasys.teku.datastructures.operations.AttesterSlashing;
+import tech.pegasys.teku.datastructures.operations.ProposerSlashing;
+import tech.pegasys.teku.datastructures.operations.SignedVoluntaryExit;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.ssz.SSZTypes.SSZList;
+import tech.pegasys.teku.ssz.SSZTypes.SSZMutableList;
+import tech.pegasys.teku.util.config.Constants;
+
+public class OperationPool<T> {
+
+  private static Map<Class<?>, Integer> maxNumberOfElementsInBlock =
+      Map.of(
+          SignedVoluntaryExit.class, Constants.MAX_VOLUNTARY_EXITS,
+          ProposerSlashing.class, Constants.MAX_PROPOSER_SLASHINGS,
+          AttesterSlashing.class, Constants.MAX_ATTESTER_SLASHINGS);
+
+  private Set<T> operations = new HashSet<>();
+  private OperationStateTransitionValidator<T> operationValidator;
+  private Class<T> clazz;
+
+  public OperationPool(Class<T> clazz, OperationStateTransitionValidator<T> operationValidator) {
+    this.clazz = clazz;
+    this.operationValidator = operationValidator;
+  }
```

Comment:
I think you could simplify this by keeping `maxNumberOfElements` as an instance var:
```suggestion
  public OperationPool(Class<T> clazz, OperationStateTransitionValidator<T> operationValidator) {
    this.operationValidator = operationValidator;
    this.maxNumberOfElements = maxNumberOfElementsInBlock.get(clazz);
  }
```

Or even cutting out `clazz` altogether:
```suggestion
  public OperationPool(OperationStateTransitionValidator<T> operationValidator, final int maxElements) {
    this.operationValidator = operationValidator;
    this.maxNumberOfElements = maxElements;
  }
```

Reply:
I like the idea of having the maxNumberOfELements get set inside the operation pool depending on class rather than someone being able to set them whatever they want. So I'll go with the first suggestion.

Reply:
Oh actually we do need the class information because we can't create the type-specific SSZMutableList if we don't.

Reply:
So unfortunately we can't make this change.